Wikipedia/Code/TalkPageViewController.swift

```diff
@@ -318,6 +325,7 @@ extension TalkPageViewController: TalkPageCellDelegate {
 
 extension TalkPageViewController: TalkPageViewModelDelegate {
     func talkPageDataDidUpdate() {
+        fakeProgressController.stop()
```

Comment:
What do you think about also calling the `stop()` method on data fetch failure? Otherwise the progress bar just hangs.

Depending on how much refactoring you're interested in, could add a new delegate method for the failure case, or just pass an error parameter in `talkPageDataDidUpdate()`. Probably the simple solution might just be passing some kind of success `BOOL` in `talkPageDataDidUpdate()`, and in the delegate implementation doing something like:

```swift
func talkPageDataDidUpdate(success: Bool) {
    fakeProgressController.stop()

    if success {
        setupHeaderView()
        talkPageView.collectionView.reloadData()
    }
}
```

Reply:
@staykids Good catch! This is updated. I ended up moving towards a completion block in  #4308 rather than `TalkPageViewModelDelegate`, due to the different followups that are needed in each context we call `fetchTalkPage`:

**Upon first load:**
- reloadData
- stopping the loading indicator
- setting up the header

**Upon posting new topic:**
- reloadData
- scrollToBottom
- present new topic alert banner

I added stopping in the load completion block whether it's success or failure. Let me know what you think, there may be a cleaner way to do it.